### PR TITLE
Issue 3: Draft PR 3+ Users

### PR DESCRIPTION
**Pull request by dillonupgradeit**
## Issue 3: Draft PR 3+ Users
Fixes: #3 

#### SEED DB:
  + Add new participant record for each user in each current conversation.
  + Calculate lastRead message for each user in each conversation and update participant data model.

#### UPDATE API
  + Versioning the API (v1, v2)
  + Conversations Endpoints => users1Id/user2Id will become to participants, otherUser will become otherUsers array (list), and add functionality to update optional conversation nickname
  + Messages Endpoints => Require conversationId and check if sender is apart of conversation before adding message using Conversation.findConversationByIdAndUserId. If conversationId null: create conversation, and add participants upon creating new conversation.

#### UTILS
  + thunkCreator/reducerFunctions => anything with recipientId will need to be updated to Array (otherUsers/participants) (ex. addNewConversation and addNewConvoToStore)
  + Will need to create new functions to update otherUsers/participants in store when adding new users to conversation.

#### COMPONENTS 
  + Sidebar => Conversations in sidebar will need to display conversation nickname by default or concatenate names of participants if no nickname exists.
  + Search for Conversations => will now use Conversation.findConversationBySearchTerm: (this checks search term against conversations nicknames or user's usernames within conversation)
  + Need to add feature for adding blank conversation in memory, then searching for and adding users to blank conversation before sending new message. 

### 2.  Migrating Production
 Assuming this project was deployed in production and had multiple interfaces like a desktop app, iOS app, android etc.:
  + The current database would need to be updated and include the new data model for participants. 
  + The participants data model would need to be seeded as explained above in Q1 (SEED DB).  
  + Versioning the API to include new functions via api/v2/[endpoints] would help tremendously, especially if the app had a desktop, iOS, or android component.
  + You could likely keep the current structure in conversations data model (user1Id and user2Id) until the other apps were updated, eventually removing user1Id and user2Id.
  
Thanks for reviewing my initial ideas on adding 3+ person conversations to the messenger chat app. I invite you to discuss and help expand this further!

### CODE REVIEW

**Comment by denis-r-Lima** [comment]
@dillonupgradeit, good job overall with this ticket.
- I liked that you keep versions for compatibility with old conversations
- on the thunkcreator what if instead of an array of users, we use a more efficient data structure like a Map or o an object
-  how are we gonna track the active conversation? right now we keep track by the otherUser

**Comment by dillonupgradeit** [comment]
> * on the thunkcreator what if instead of an array of users, we use a more efficient data structure like a Map or o an object

I'm glad you brought up this point, and believe the object could be a very good data structure to handle participants particularly when we bring lookups and JSON in to the mix. In my experience, an object doesn't really present much more (sometimes less) efficiency than an array objects containing less than say 100 objects. And I doubt very many conversations would contain that many participants. However it really depends on the type of loops and functions we end up using on this array/object of users. If we are simply looping through and displaying the user on front end, array might be a better option. But if we are doing lookups/filtering or looping through to find a particular user, objects would likely be more efficient. 

> * how are we gonna track the active conversation? right now we keep track by the otherUser

Ah! Another great point. This is something I thought about quite a bit while working on the other issues, but failed to include it in Issue_3. I thought of two different methods to keep track of active conversation. One would be to create a record of a conversation in the db -> returning the conversationID as soon as a user started a blank conversation (even before adding participants), and keep track of the conversationId in active conversation. This would require we delete the record if no initial message was ever sent (not very elegant). Another method would be to create a random temporary nickname or even temporary conversationId (ex. nickname = "Wild Wombats" OR conversationId = TMP-[userid]-1) for the conversation until it was created in the db, and in activeConversation if no conversastionId exists, use the temporary nickname/temporary conversationId.

File: server/db/models/index.js [hunk 1]
```diff
@@ -1,17 +1,20 @@
...
 // associations
 
-User.hasMany(Conversation);
-Conversation.belongsTo(User, { as: "user1" });
-Conversation.belongsTo(User, { as: "user2" });
-Message.belongsTo(Conversation);
+Conversation.belongsToMany(User, { through: Participant });
+Conversation.belongsToMany(User, {through: Participant, as: 'filterUser'});
 Conversation.hasMany(Message);
+Conversation.hasMany(Participant);
```
Comment (FilipeGorodscy) [inline server/db/models/index.js:11]:
```suggestion
Conversation.belongsToMany(User, {through: Participant});
Users.belongsToMany(Conversation, {through: Participant});
Conversation.hasMany(Message);
```

File: server/db/models/participant.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  const participant = await Participant.findOne({
+    where: {
+      conversationId,
+      userId,
+    }
+  });
+
+  // return participant or null if it doesn't exist
+  return participant;
+};
```
Comment (FilipeGorodscy) [inline server/db/models/participant.js:16]: You could query the conversation table instead :)

File: server/db/models/participant.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// find conversation given conversationId
+Participant.findActiveParticipantsByConversationId = async function (conversationId) {
+  const participants = await Participant.findAll({
+    where: {
+      conversationId,
+    }
+  });
+
+  // return conversation or null if it doesn't exist
+  return participants;
```
Comment (FilipeGorodscy) [inline server/db/models/participant.js:27]: Same as above

**Review by FilipeGorodscy — APPROVED** [review]
Good job! The only thing is that the Many-To-Many relationship between Users and Conversations could have been simplified. This PR is approved :)

File: server/db/models/index.js (hunk 1 above)
Comment (dillonupgradeit) [reply server/db/models/index.js:11]: Oooh. You're so very correct sir. Very elegant.

**Pull request closed without merging**